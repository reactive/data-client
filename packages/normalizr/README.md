# Normalizr Client

[![CircleCI](https://circleci.com/gh/reactive/data-client/tree/master.svg?style=shield)](https://circleci.com/gh/data-clientdata-clients)
[![Coverage Status](https://img.shields.io/codecov/c/gh/reactive/data-client/master.svg?style=flat-square)](https://app.codecov.io/gh/data-clientdata-clients?branch=master)
[![npm downloads](https://img.shields.io/npm/dt/@data-client/normalizr.svg?style=flat-square)](https://www.npmjs.com/package/@data-client/normalizr)
[![bundle size](https://img.shields.io/bundlephobia/minzip/@data-client/normalizr?style=flat-square)](https://bundlephobia.com/result?p=@data-client/normalizr)
[![npm version](https://img.shields.io/npm/v/@data-client/normalizr.svg?style=flat-square)](https://www.npmjs.com/package/data-clients/normalizr) [![npm downloads](https://img.shields.io/npm/dm/@data-client/normalizr.svg?style=flat-square)](https://www.npmjs.com/packagdata-clientoks/normalizr)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](http://makeapullrequest.com)

## Install

Install from the NPM repository using yarn or npm:

```shell
yarn add @data-client/normalizr
```

```shell
npm install --save @data-client/normalizr
```

## Motivation

Many APIs, public or not, return JSON data that has deeply nested objects. Using data in this kind of structure is often very difficult for JavaScript applications, especially those using [Flux](http://facebook.github.io/flux/) or [Redux](http://redux.js.org/).

## Solution

Normalizr is a small, but powerful utility for taking JSON with a schema definition and returning nested entities with their IDs, gathered in dictionaries.

## Documentation

- [Introduction](https://dataclient.io/docs/concepts/normalization)
- [Quick Start](https://dataclient.io/rest/api/schema)
- [API](https://dataclient.io/rest/api/Entity)
  - [normalize](./docs/api.md#normalizedata-schema)
  - [denormalize](./docs/api.md#denormalizeinput-schema-entities)
  - [schema](https://dataclient.io/rest/api/schema)

## Examples

- [Normalizing GitHub Issues](../../examples/normalizr-github)
- [Relational Data](../../examples/normalizr-relationships)
- [Interactive Redux](../../examples/normalizr-redux)
- [Benchmarks](../../examples/benchmark)
  - [Charts](https://reactive.github.io/data-client/dev/bench/)

## React Demos

- [Todo Demo](https://stackblitz.com/github/reactive/data-client/tree/master/examples/todo-app?file=src%2Fpages%2FHome%2FTodoList.tsx)
- [Github Demo](https://stackblitz.com/github/reactive/data-client/tree/master/examples/github-app?file=src%2Fpages%2FIssueList.tsx)
- [NextJS SSR Demo](https://stackblitz.com/github/reactive/data-client/tree/master/examples/nextjs?file=components%2Ftodo%2FTodoList.tsx)

## Quick Start

Consider a typical blog post. The API response for a single post might look something like this:

```json
{
  "id": "123",
  "author": {
    "id": "1",
    "name": "Paul"
  },
  "title": "My awesome blog post",
  "comments": [
    {
      "id": "324",
      "createdAt": "2013-05-29T00:00:00-04:00",
      "commenter": {
        "id": "2",
        "name": "Nicole"
      }
    },
    {
      "id": "544",
      "createdAt": "2013-05-30T00:00:00-04:00",
      "commenter": {
        "id": "1",
        "name": "Paul"
      }
    }
  ]
}
```

We have two [nested](https://dataclient.io/rest/guides/relational-data) [entity](https://dataclient.io/rest/api/Entity) types within our `article`: `users` and `comments`. Using various [schema](https://dataclient.io/rest/api/Entity#schema), we can normalize all three entity types down:

```js
import { schema, Entity } from '@data-client/endpoint';
import { Temporal } from '@js-temporal/polyfill';

// Define a users schema
class User extends Entity {}

// Define your comments schema
class Comment extends Entity {
  static schema = {
    commenter: User,
    createdAt: Temporal.Instant.from,
  };
}

// Define your article
class Article extends Entity {
  static schema = {
    author: User,
    comments: [Comment],
  };
}
```

### Normalize

```js
import { normalize } from '@data-client/normalizr';

const args = [{ id: '123' }];
const normalizedData = normalize(Article, originalData, args);
```

Now, `normalizedData` will create a single serializable source of truth for all entities:

```js
{
  result: "123",
  entities: {
    articles: {
      "123": {
        id: "123",
        author: "1",
        title: "My awesome blog post",
        comments: [ "324", "544" ]
      }
    },
    users: {
      "1": { "id": "1", "name": "Paul" },
      "2": { "id": "2", "name": "Nicole" }
    },
    comments: {
      "324": {
        id: "324",
        createdAt: "2013-05-29T00:00:00-04:00",
        commenter: "2"
      },
      "544": {
        id: "544",
        createdAt: "2013-05-30T00:00:00-04:00",
        commenter: "1"
      }
    }
  },
}
```

`normalizedData` can be placed in any flux store as the single source of truth for this data.

### Denormalize

Accessing the store can then be done using flux `selectors` by `denormalizing`:

```js
import { denormalize } from '@data-client/normalizr';

const denormalizedData = denormalize(
  Article,
  normalizedData.result,
  normalizedData.entities,
  args,
);
```

Now, `denormalizedData` will instantiate the classes, ensuring all instances of the same member (like `Paul`) are referentially equal:

```js
Article {
  id: '123',
  title: 'My awesome blog post',
  author: User { id: '1', name: 'Paul' },
  comments: [
    Comment {
      id: '324',
      createdAt: Instant [Temporal.Instant] {},
      commenter: [User { id: '2', name: 'Nicole' }]
    },
    Comment {
      id: '544',
      createdAt: Instant [Temporal.Instant] {},
      commenter: [User { id: '1', name: 'Paul' }]
    }
  ]
}
```

### MemoCache

`MemoCache` is a singleton that can be used to maintain referential equality between calls as well
as potentially improved performance by 2000%. All three methods are memoized.

```js
import { MemoCache } from '@data-client/normalizr';

// you can construct a new memo anytime you want to reset the cache
const memo = new MemoCache();

const { data, paths } = memo.denormalize(schema, input, state.entities, args);

({ data, paths } = memo.query(schema, args, state));

function query(schema, args, state, key) {
  const queryKey = memo.buildQueryKey(
    schema,
    args,
    state,
    key,
  );
  const { data } = this.denormalize(schema, queryKey, state.entities, args);
  return typeof data === 'symbol' ? undefined : (data as any);
}
```

`memo.denormalize()` is just like denormalize() above but includes `paths` as part of the return value. `paths`
is an Array of paths of all entities included in the result.

`memo.query()` allows denormalizing without a normalized input. See [Queryable](https://dataclient.io/rest/api/schema#queryable) for more info.

`memo.buildQueryKey()` builds the input used to denormalize for `query()`. This is exposed
to allow greater flexibility in its usage.

## Schemas

Available from [@data-client/endpoint](https://www.npmjs.com/package/@data-client/endpoint)

<table>
<thead>
<tr>
<th>Data Type</th>
<th>Mutable</th>
<th>Schema</th>
<th>Description</th>
<th><a href="https://dataclient.io/rest/api/schema#queryable">Queryable</a></th>
</tr>
</thead>
<tbody><tr>
<td rowSpan="4"><a href="https://en.wikipedia.org/wiki/Object_(computer_science)">Object</a></td>
<td align="center">âœ…</td>
<td><a href="https://dataclient.io/rest/api/Entity">Entity</a>, <a href="https://dataclient.io/rest/api/EntityMixin">EntityMixin</a></td>
<td>single <em>unique</em> object</td>
<td align="center">âœ…</td>
</tr>
<tr>
<td align="center">âœ…</td>
<td><a href="https://dataclient.io/rest/api/Union">Union(Entity)</a></td>
<td>polymorphic objects (<code>A | B</code>)</td>
<td align="center">âœ…</td>
</tr>
<tr>
<td align="center">ðŸ›‘</td>
<td><a href="https://dataclient.io/rest/api/Object">Object</a></td>
<td>statically known keys</td>
<td align="center">ðŸ›‘</td>
</tr>
<tr>
<td align="center"></td>
<td><a href="https://dataclient.io/rest/api/Invalidate">Invalidate(Entity)</a></td>
<td><a href="https://dataclient.io/docs/concepts/expiry-policy#invalidate-entity">delete an entity</a></td>
<td align="center">ðŸ›‘</td>
</tr>
<tr>
<td rowSpan="3"><a href="https://en.wikipedia.org/wiki/List_(abstract_data_type)">List</a></td>
<td align="center">âœ…</td>
<td><a href="https://dataclient.io/rest/api/Collection">Collection(Array)</a></td>
<td>growable lists</td>
<td align="center">âœ…</td>
</tr>
<tr>
<td align="center">ðŸ›‘</td>
<td><a href="https://dataclient.io/rest/api/Array">Array</a></td>
<td>immutable lists</td>
<td align="center">ðŸ›‘</td>
</tr>
<tr>
<td align="center"> </td>
<td><a href="https://dataclient.io/rest/api/All">All</a></td>
<td>list of all entities of a kind</td>
<td align="center">âœ…</td>
</tr>
<tr>
<td rowSpan="2"><a href="https://en.wikipedia.org/wiki/Associative_array">Map</a></td>
<td align="center">âœ…</td>
<td><a href="https://dataclient.io/rest/api/Collection">Collection(Values)</a></td>
<td>growable maps</td>
<td align="center">âœ…</td>
</tr>
<tr>
<td align="center">ðŸ›‘</td>
<td><a href="https://dataclient.io/rest/api/Values">Values</a></td>
<td>immutable maps</td>
<td align="center">ðŸ›‘</td>
</tr>
<tr>
<td>any</td>
<td align="center"></td>
<td><a href="https://dataclient.io/rest/api/Query">Query(Queryable)</a></td>
<td>memoized custom transforms</td>
<td align="center">âœ…</td>
</tr>
</tbody></table>

## Benchmarks

[Performance compared](https://github.com/reactive/data-client/blob/master/examples/benchmark/README.md) to normalizr package (higher is better):

|                     | no cache | with cache |
| ------------------- | -------- | ---------- |
| normalize (long)    | 120%     | 120%       |
| denormalize (long)  | 158%     | 1,250%     |
| denormalize (short) | 676%     | 2,650%     |

[View benchmark](https://github.com/reactive/data-client/blob/master/examples/benchmark)

## Credits

Normalizr Client is based on [Normalizr](https://github.com/paularmstrong/normalizr) - originally created by [Dan Abramov](http://github.com/gaearon) and inspired by a conversation with [Jing Chen](https://twitter.com/jingc). Since v3, it was completely rewritten and maintained by [Paul Armstrong](https://twitter.com/paularmstrong).

Normalizr Client was rewritten and maintained by Normalizr contributor [Nathaniel Tucker](https://twitter.com/npinp). It has also received much help, enthusiasm, and contributions from [community members](https://github.com/ntucker/normalizr/graphs/contributors).
