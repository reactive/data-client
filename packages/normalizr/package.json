{
  "name": "@data-client/normalizr",
  "version": "0.14.17",
  "description": "Normalizes and denormalizes JSON according to schema for Redux and Flux applications",
  "homepage": "https://dataclient.io/docs/concepts/normalization",
  "keywords": [
    "normalize",
    "denormalize",
    "flux",
    "redux",
    "data",
    "model",
    "cache",
    "api",
    "memoize",
    "react-native",
    "ios",
    "android",
    "web",
    "schema",
    "typescript",
    "validation",
    "relational",
    "reactive",
    "json"
  ],
  "sideEffects": false,
  "main": "dist/normalizr.js",
  "react-native": "legacy/index.js",
  "module": "legacy/index.js",
  "types": "lib/index.d.ts",
  "typesVersions": {
    ">=5.4": {
      "": [
        "lib/index.d.ts"
      ],
      "*": [
        "lib/index.d.ts"
      ]
    },
    ">=4.1": {
      "": [
        "ts4.1/index.d.ts"
      ],
      "*": [
        "ts4.1/index.d.ts"
      ]
    },
    ">=4.0": {
      "": [
        "ts4.0/index.d.ts"
      ],
      "*": [
        "ts4.0/index.d.ts"
      ]
    },
    ">=3.4": {
      "": [
        "ts3.4/index.d.ts"
      ],
      "*": [
        "ts3.4/index.d.ts"
      ]
    }
  },
  "exports": {
    ".": {
      "module": "./lib/index.js",
      "import": "./node.mjs",
      "require": "./dist/normalizr.js",
      "default": "./lib/index.js"
    },
    "./package.json": "./package.json"
  },
  "type": "module",
  "engines": {
    "node": "^12.17 || ^13.7 || >=14"
  },
  "files": [
    "src",
    "dist",
    "lib",
    "node.mjs",
    "legacy",
    "ts4.1",
    "ts4.0",
    "ts3.4",
    "LICENSE",
    "README.md"
  ],
  "scripts": {
    "build": "run build:lib && run build:legacy:lib && run build:bundle",
    "build:js:development": "BROWSERSLIST_ENV=legacy NODE_ENV=development yarn g:rollup",
    "build:js:production": "BROWSERSLIST_ENV=legacy NODE_ENV=production yarn g:rollup",
    "build:js:node": "BROWSERSLIST_ENV=node12 NODE_ENV=development yarn g:rollup && echo '{\"type\":\"commonjs\"}' > dist/package.json",
    "build:lib": "NODE_ENV=production BROWSERSLIST_ENV='2020' yarn g:babel --out-dir lib",
    "build:legacy:lib": "NODE_ENV=production BROWSERSLIST_ENV='2018' yarn g:babel --out-dir legacy",
    "build:bundle": "yarn g:runs build:js:\\*",
    "build:clean": "yarn g:clean ts3.4 ts4.0 ts4.1",
    "build:legacy-types": "yarn g:legacy-types 4.1 4.0 3.4",
    "prepublishOnly": "run build"
  },
  "author": "Nathaniel Tucker <me@ntucker.me> (https://github.com/ntucker)",
  "funding": "https://github.com/sponsors/ntucker",
  "contributors": [
    "Paul Armstrong",
    "Dan Abramov"
  ],
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com:reactive/data-client.git",
    "directory": "packages/normalizr"
  },
  "bugs": {
    "url": "https://github.com/reactive/data-client/issues"
  },
  "dependencies": {
    "@babel/runtime-corejs3": "^7.26.0",
    "core-js-pure": "^3.40.0"
  },
  "devDependencies": {
    "@anansi/browserslist-config": "^1.4.2",
    "@data-client/endpoint": "workspace:*",
    "@js-temporal/polyfill": "^0.4.4",
    "@types/jest": "^29.5.14",
    "@types/node": "^22.0.0",
    "immutable": "4.2.2",
    "rollup-plugins": "workspace:*"
  }
}
