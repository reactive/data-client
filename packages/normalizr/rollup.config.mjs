import {
  babel,
  commonjs,
  filesize,
  dts,
  resolve,
  terser,
  onwarn,
} from 'rollup-plugins';

import pkg from './package.json' with { type: 'json' };

const dependencies = Object.keys(pkg.dependencies).filter(
  dep => !['@babel/runtime'].includes(dep),
);

const name = 'normalizr';

const isProduction = process.env.NODE_ENV === 'production';

const destBase = 'dist/normalizr';
const destExtension = `${isProduction ? '.min' : ''}.js`;
const extensions = ['.js', '.ts', '.tsx', '.mjs', '.json', '.node'];

function snakeCase(id) {
  return id.replace(/(-|\/)/g, '_').replace(/@/g, '');
}

function isExternal(id) {
  return dependencies.some(dep => dep === id || id.startsWith(dep));
}

const configs = [];

if (process.env.BROWSERSLIST_ENV !== 'node12') {
  if (process.env.NODE_ENV === 'production') {
    configs.push({
      input: 'src/index.ts',
      output: [{ file: `${destBase}.es${destExtension}`, format: 'es' }],
      external: id => id === '..' || isExternal(id),
      onwarn,
      plugins: [
        babel({
          exclude: ['node_modules/**', '/**__tests__/**'],
          extensions,
          babelHelpers: 'runtime',
          rootMode: 'upward',
        }),
        resolve({ extensions }),
        commonjs({ extensions }),
        isProduction && terser(),
        filesize(),
      ].filter(Boolean),
    });
  } else {
    configs.push({
      input: 'src/index.ts',
      output: [
        {
          file: `${destBase}.umd${destExtension}`,
          format: 'umd',
          name: snakeCase(name),
        },
        {
          file: `${destBase}.amd${destExtension}`,
          format: 'amd',
          name: snakeCase(name),
        },
        {
          file: `${destBase}.browser${destExtension}`,
          format: 'iife',
          name: snakeCase(name),
        },
      ],
      onwarn,
      plugins: [
        babel({
          exclude: ['node_modules/**', '/**__tests__/**'],
          extensions,
          babelHelpers: 'runtime',
          rootMode: 'upward',
          caller: { polyfillMethod: false },
        }),
        resolve({ extensions }),
        commonjs({ extensions }),
        isProduction && terser(),
        filesize(),
      ].filter(Boolean),
    });
    configs.push({
      input: './lib/index.d.ts',
      output: [{ file: 'index.d.ts', format: 'es' }],
      plugins: [dts()],
    });
  }
} else {
  // node-friendly commonjs build
  configs.push({
    input: 'src/index.ts',
    output: [{ file: `${destBase}${destExtension}`, format: 'cjs' }],
    external: id => id === '..' || isExternal(id),
    onwarn,
    plugins: [
      babel({
        exclude: ['node_modules/**', '/**__tests__/**'],
        extensions,
        babelHelpers: 'runtime',
        rootMode: 'upward',
      }),
      resolve({ extensions }),
      commonjs({ extensions }),
      isProduction && terser(),
      filesize(),
    ].filter(Boolean),
  });
}
export default configs;
