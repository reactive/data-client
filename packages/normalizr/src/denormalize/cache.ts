import type { EntityInterface } from '../interface.js';
import { EntityPath } from '../types.js';
import type { INVALID } from './symbol.js';

export default interface Cache {
  getEntity(
    pk: string,
    schema: EntityInterface,
    entity: any,
    computeValue: (localCacheKey: Map<string, any>) => void,
  ): object | undefined | typeof INVALID;
  getResults(
    input: any,
    cachable: boolean,
    computeValue: () => any,
  ): {
    data: any;
    paths: EntityPath[];
  };
}
