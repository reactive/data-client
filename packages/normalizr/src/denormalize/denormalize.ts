import { getEntities } from './getEntities.js';
import LocalCache from './localCache.js';
import getUnvisit from './unvisit.js';
import type { Schema } from '../interface.js';
import { isImmutable } from '../schemas/ImmutableUtils.js';
import type { DenormalizeNullable } from '../types.js';
import type { INVALID } from './symbol.js';

export function denormalize<S extends Schema>(
  schema: S | undefined,
  input: any,
  entities: any,
  args: readonly any[] = [],
): DenormalizeNullable<S> | typeof INVALID {
  // undefined means don't do anything
  if (schema === undefined || input === undefined) {
    return input as any;
  }

  return getUnvisit(
    getEntities(entities),
    new LocalCache(),
    args,
    isImmutable(entities),
  )(schema, input).data;
}
