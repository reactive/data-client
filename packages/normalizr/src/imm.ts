export { MemoPolicy } from './memo/Policy.imm.js';
export { denormalize } from './denormalize/denormalize.imm.js';
