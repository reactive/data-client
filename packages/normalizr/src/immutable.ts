export { DelegateImmutable } from './memo/Delegate.immutable.js';
