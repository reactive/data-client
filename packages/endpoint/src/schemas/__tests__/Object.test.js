// eslint-env jest
import {
  normalize,
  denormalize as plainDenormalize,
} from '@data-client/normalizr';
import { denormalize as immDenormalize } from '@data-client/normalizr/imm';
import { Temporal } from '@js-temporal/polyfill';
import { IDEntity } from '__tests__/new';
import { fromJS } from 'immutable';

import { fromJSEntities } from './denormalize';
import { schema } from '../../';
import Entity from '../Entity';

let dateSpy;
beforeAll(() => {
  dateSpy = jest

    .spyOn(global.Date, 'now')
    .mockImplementation(() => new Date('2019-05-14T11:01:58.135Z').valueOf());
});
afterAll(() => {
  dateSpy.mockRestore();
});

describe(`${schema.Object.name} normalization`, () => {
  test('normalizes an object', () => {
    class User extends IDEntity {}
    const object = new schema.Object({
      user: User,
    });
    expect(normalize(object, { user: { id: '1' } })).toMatchSnapshot();
  });

  test(`normalizes plain objects as shorthand for ${schema.Object.name}`, () => {
    class User extends IDEntity {}
    expect(normalize({ user: User }, { user: { id: '1' } })).toMatchSnapshot();
  });

  test('filters out undefined and null values', () => {
    class User extends Entity {}
    const users = new schema.Object({ foo: User, bar: User, baz: User });
    const oldenv = process.env.NODE_ENV;
    process.env.NODE_ENV = 'production';
    expect(
      normalize(users, { foo: undefined, bar: { id: '1' } }),
    ).toMatchSnapshot();
    process.env.NODE_ENV = oldenv;
  });

  it('should deserialize Date', () => {
    class User extends IDEntity {}
    const WithOptional = new schema.Object({
      user: User,
      nextPage: '',
      createdAt: Temporal.Instant.from,
    });
    const normalized = normalize(WithOptional, {
      user: { id: '5' },
      nextPage: 'blob',
      createdAt: '2020-06-07T02:00:15.000Z',
    });
    expect(normalized.result.createdAt).toBe(normalized.result.createdAt);
    expect(typeof normalized.result.createdAt).toBe('string');
    expect(normalized).toMatchSnapshot();
  });

  it('should pass over when Date not provided', () => {
    class User extends IDEntity {}
    const WithOptional = new schema.Object({
      user: User,
      nextPage: '',
      createdAt: Temporal.Instant.from,
    });
    const normalized = normalize(WithOptional, {
      user: { id: '5' },
      nextPage: 'blob',
    });
    expect(normalized.result.createdAt).toBeUndefined();
    expect(normalized).toMatchSnapshot();
  });
});

describe(`${schema.Object.name} denormalization`, () => {
  test('denormalizes an object', () => {
    class User extends IDEntity {}
    const object = new schema.Object({
      user: User,
    });
    const entities = {
      User: {
        1: { id: '1', name: 'Nacho' },
      },
    };
    expect(plainDenormalize(object, { user: '1' }, entities)).toMatchSnapshot();
    expect(
      immDenormalize(object, { user: '1' }, fromJSEntities(entities)),
    ).toMatchSnapshot();
    expect(
      immDenormalize(object, fromJS({ user: '1' }), fromJSEntities(entities)),
    ).toMatchSnapshot();
  });

  test('denormalizes an object with plain members', () => {
    class User extends IDEntity {}
    const object = new schema.Object({
      user: User,
      plain: '',
    });
    const entities = {
      User: {
        1: { id: '1', name: 'Nacho' },
      },
    };
    expect(plainDenormalize(object, { user: '1' }, entities)).toMatchSnapshot();
    expect(
      immDenormalize(object, { user: '1' }, fromJSEntities(entities)),
    ).toMatchSnapshot();
    expect(
      immDenormalize(object, fromJS({ user: '1' }), fromJSEntities(entities)),
    ).toMatchSnapshot();
  });

  test('should have found = true with null member even when schema has nested entity', () => {
    class User extends IDEntity {}
    const object = new schema.Object({
      item: new schema.Object({
        user: User,
      }),
    });
    const entities = {
      User: {
        1: { id: '1', name: 'Nacho' },
      },
    };
    let value = plainDenormalize(object, { item: null }, entities);
    expect(value).toMatchSnapshot();
    value = immDenormalize(object, { item: null }, fromJSEntities(entities));
    expect(value).toMatchSnapshot();
    value = immDenormalize(
      object,
      fromJS({ item: null }),
      fromJSEntities(entities),
    );
    expect(value).toMatchSnapshot();
  });

  test('denormalizes plain object shorthand', () => {
    class User extends IDEntity {}
    const entities = {
      User: {
        1: { id: '1', name: 'Jane' },
      },
    };
    expect(
      plainDenormalize(
        new schema.Object({ user: User, tacos: {} }),
        { user: '1' },
        entities,
      ),
    ).toMatchSnapshot();
    expect(
      immDenormalize(
        new schema.Object({ user: User, tacos: {} }),
        { user: '1' },
        fromJSEntities(entities),
      ),
    ).toMatchSnapshot();
    expect(
      immDenormalize(
        new schema.Object({ user: User, tacos: {} }),
        fromJS({ user: '1' }),
        fromJSEntities(entities),
      ),
    ).toMatchSnapshot();

    expect(
      plainDenormalize(
        new schema.Object({ user: User, tacos: {} }),
        { user: '1', tacos: {} },
        entities,
      ),
    ).toMatchSnapshot();
    expect(
      immDenormalize(
        new schema.Object({ user: User, tacos: {} }),
        { user: '1', tacos: {} },
        fromJSEntities(entities),
      ),
    ).toMatchSnapshot();
    expect(
      immDenormalize(
        new schema.Object({ user: User, tacos: {} }),
        fromJS({ user: '1', tacos: {} }),
        fromJSEntities(entities),
      ),
    ).toMatchSnapshot();
  });

  test('denormalizes an object that contains a property representing a an object with an id of zero', () => {
    class User extends IDEntity {}
    const object = new schema.Object({
      user: User,
    });
    const entities = {
      User: {
        0: { id: '0', name: 'Chancho' },
      },
    };
    expect(plainDenormalize(object, { user: '0' }, entities)).toMatchSnapshot();
    expect(
      immDenormalize(object, { user: '0' }, fromJSEntities(entities)),
    ).toMatchSnapshot();
    expect(
      immDenormalize(object, fromJS({ user: '0' }), fromJSEntities(entities)),
    ).toMatchSnapshot();
  });
});
