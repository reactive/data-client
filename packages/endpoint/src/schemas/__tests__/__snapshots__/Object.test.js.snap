// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectSchema denormalization denormalizes an object 1`] = `
{
  "user": User {
    "id": "1",
    "name": "Nacho",
  },
}
`;

exports[`ObjectSchema denormalization denormalizes an object 2`] = `
{
  "user": User {
    "id": "1",
    "name": "Nacho",
  },
}
`;

exports[`ObjectSchema denormalization denormalizes an object 3`] = `
Immutable.Map {
  "user": User {
    "id": "1",
    "name": "Nacho",
  },
}
`;

exports[`ObjectSchema denormalization denormalizes an object that contains a property representing a an object with an id of zero 1`] = `
{
  "user": User {
    "id": "0",
    "name": "Chancho",
  },
}
`;

exports[`ObjectSchema denormalization denormalizes an object that contains a property representing a an object with an id of zero 2`] = `
{
  "user": User {
    "id": "0",
    "name": "Chancho",
  },
}
`;

exports[`ObjectSchema denormalization denormalizes an object that contains a property representing a an object with an id of zero 3`] = `
Immutable.Map {
  "user": User {
    "id": "0",
    "name": "Chancho",
  },
}
`;

exports[`ObjectSchema denormalization denormalizes an object with plain members 1`] = `
{
  "user": User {
    "id": "1",
    "name": "Nacho",
  },
}
`;

exports[`ObjectSchema denormalization denormalizes an object with plain members 2`] = `
{
  "user": User {
    "id": "1",
    "name": "Nacho",
  },
}
`;

exports[`ObjectSchema denormalization denormalizes an object with plain members 3`] = `
Immutable.Map {
  "user": User {
    "id": "1",
    "name": "Nacho",
  },
}
`;

exports[`ObjectSchema denormalization denormalizes plain object shorthand 1`] = `
{
  "user": User {
    "id": "1",
    "name": "Jane",
  },
}
`;

exports[`ObjectSchema denormalization denormalizes plain object shorthand 2`] = `
{
  "user": User {
    "id": "1",
    "name": "Jane",
  },
}
`;

exports[`ObjectSchema denormalization denormalizes plain object shorthand 3`] = `
Immutable.Map {
  "user": User {
    "id": "1",
    "name": "Jane",
  },
}
`;

exports[`ObjectSchema denormalization denormalizes plain object shorthand 4`] = `
{
  "tacos": {},
  "user": User {
    "id": "1",
    "name": "Jane",
  },
}
`;

exports[`ObjectSchema denormalization denormalizes plain object shorthand 5`] = `
{
  "tacos": {},
  "user": User {
    "id": "1",
    "name": "Jane",
  },
}
`;

exports[`ObjectSchema denormalization denormalizes plain object shorthand 6`] = `
Immutable.Map {
  "user": User {
    "id": "1",
    "name": "Jane",
  },
  "tacos": Immutable.Map {},
}
`;

exports[`ObjectSchema denormalization should have found = true with null member even when schema has nested entity 1`] = `
{
  "item": null,
}
`;

exports[`ObjectSchema denormalization should have found = true with null member even when schema has nested entity 2`] = `
{
  "item": null,
}
`;

exports[`ObjectSchema denormalization should have found = true with null member even when schema has nested entity 3`] = `
Immutable.Map {
  "item": null,
}
`;

exports[`ObjectSchema normalization filters out undefined and null values 1`] = `
{
  "entities": {
    "User": {
      "1": {
        "id": "1",
      },
    },
  },
  "entitiesMeta": {
    "User": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": {
    "bar": "1",
  },
}
`;

exports[`ObjectSchema normalization normalizes an object 1`] = `
{
  "entities": {
    "User": {
      "1": {
        "id": "1",
      },
    },
  },
  "entitiesMeta": {
    "User": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": {
    "user": "1",
  },
}
`;

exports[`ObjectSchema normalization normalizes plain objects as shorthand for ObjectSchema 1`] = `
{
  "entities": {
    "User": {
      "1": {
        "id": "1",
      },
    },
  },
  "entitiesMeta": {
    "User": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": {
    "user": "1",
  },
}
`;

exports[`ObjectSchema normalization should deserialize Date 1`] = `
{
  "entities": {
    "User": {
      "5": {
        "id": "5",
      },
    },
  },
  "entitiesMeta": {
    "User": {
      "5": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": {
    "createdAt": "2020-06-07T02:00:15.000Z",
    "nextPage": "blob",
    "user": "5",
  },
}
`;

exports[`ObjectSchema normalization should pass over when Date not provided 1`] = `
{
  "entities": {
    "User": {
      "5": {
        "id": "5",
      },
    },
  },
  "entitiesMeta": {
    "User": {
      "5": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": {
    "nextPage": "blob",
    "user": "5",
  },
}
`;
